File: pi.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
pi.py
=======

Process information for specific user.

Copyright (C) 2012 Yi-Xin Liu

"""

import argparse
import socket
import os.path
import datetime

import rpyc

parser = argparse.ArgumentParser(description='pi.py options.\nCopyright'
                                             '(C) 2012 Yi-Xin Liu'
                                             ' <liuyxpp@gmail.com>')

parser.add_argument('-u', '--user',
                    default='lyx',
                    help='Username.')

parser.add_argument('-n', '--nodes',
                    default=[],
                    nargs='+',
                    help='A list of nodes to be looked up.')

parser.add_argument('-f', '--fail',
                    action='store_true',
                    help='Whether or not to show failed infomation.')

args = parser.parse_args()

__version__ = 0.1

time_format = '%Y-%m-%d %H:%M:%S'

all_nodes = ['c0101', 'c0102', 'c0103', 'c0104',
             'c0105', 'c0106', 'c0107', 'c0108', 'c0109',
             'c0110', 'c0111', 'c0112', 'c0113', 'c0114',
             'c0115', 'c0116', 'c0117', 'c0118', 'c0119',
             'c0120', 'c0121', 'c0122', 'c0123', 'c0124',
            ]


def print_user_process(node, process):
    name = process.name
    ptime = process.create_time
    raw_create_time = datetime.datetime.fromtimestamp(ptime)
    create_time = raw_create_time.strftime(time_format)
    #if name == 'python':
    #    name = os.path.basename(process.getcwd())
    print '\033[35;49m\t', node,
    print '\033[m\t', name,
    print '\033[32;49m\t', process.pid,
    print '\033[m\t', process.get_cpu_percent(),
    print '\033[36;49m\t', create_time,
    print '\033[m'


def find_user_process(node, conn):
    np = 0
    for p in conn.modules.psutil.process_iter():
        if p.username == args.user:
            if p.name == 'bash' or p.name == 'rpyc_classic.py':
                continue
            if p.get_cpu_percent() < 3.0:
                continue
            print_user_process(node, p)
            np += 1
    return np


def run():
    if args.nodes is None:
        nodes = ['localhost']
    elif 'all' in args.nodes:
        nodes = all_nodes
    else:
        nodes = args.nodes

    print 'The processes of user \033[33;49m[', args.user,
    print ']\033[m are:'
    print
    print '\033[35;49m\t NODE\033[m\tPROGRAM\t\033[32;49m\tPID',
    print '\033[m\tCPU%\033[36;49m\tTime'
    print '\033[m'

    failed_nodes = {}  # node: failed message
    num_procs = {}  # node: number of processes
    total_procs = 0  # total number of processes for the node list
    for node in nodes:
        try:
            conn = rpyc.classic.connect(node)
        except socket.timeout, e:
            failed_nodes[node] = e
            continue
        except socket.error, e:
            failed_nodes[node] = e
            continue

        np = find_user_process(node, conn)
        num_procs[node] = np
        total_procs += np

    print
    print 'Total processes of\033[33;49m', args.user,
    print '\033[m: \033[32;49m[', total_procs,
    print ']\033[m.'
    if total_procs > 0:
        print 'Of which:'
        for node in sorted(num_procs.iterkeys()):
            np = num_procs[node]
            if np > 0:
                print '\033[35;49m\t', node,
                print '\033[m\033[32;49m\t', np,
                print '\033[m'

    if args.fail and len(failed_nodes) > 0:
        print 'Failed to connect to following nodes:'
        for node in sorted(failed_nodes.iterkeys()):
            msg = failed_nodes[node]
            print '\033[35;49m\t', node,
            print '\033[mbecause of\033[0;31m', msg, '\033[m.'
    print


if __name__ == '__main__':
    run()

File: pi.pl
#!/usr/bin/perl
# pi - process infomation for specific user
#
# Author: Jie Gao @ Fudan Univ.
#
# Revision: Yi-Xin Liu @ Fudan Univ.
# Since 9/23/2010
#
use Getopt::Std;

my %opt;
getopts('u:v',\%opt);
if(defined $opt{'v'}){
	print "This is pi, version 9/23/2010.\n\n";
	print "pi is a perl script for listing process infomation for the user specified by option -u.\n\n";
	print "Author: Yi-Xin Liu.\n";
	print "Contact: liuyxpp at gmail.com.\n";
	exit(0);
}
my $user="lyx"; # default user
$user=$opt{'u'} if defined $opt{'u'};

my $count=1;
my $total=0;
my %total_node;
print "The processes of user \033[1;33;49m[$user]\033[m are:\n\n";
print "\t\033[1;35;49mNODE\033[m\tPROGRAM\t\t\033[1;32;49mPID\033[m\tCPU%\n";
while($count++<16){
	my $node;
	if($count<10){
		$node = "c010$count";
	}
	else{
		$node = "c01$count";
	}
	my $rsh="rsh $node ";
	my $ps_cmd="ps -ef |grep $user |grep -v grep |grep -v bash |tr -s \" \" \" \"";
#	my @ps=`$rsh $ps_cmd`;
	my $ps_cmd_pid=$ps_cmd."|cut -f2 -d \" \"";
	my @pid=split "\n", `$rsh $ps_cmd_pid`;
	my $size=@pid;
	my $ps_cmd_cpu=$ps_cmd."|cut -f4 -d \" \"";
	my @cpu=split "\n", `$rsh $ps_cmd_cpu`;
	my $ps_cmd_exe=$ps_cmd."|cut -f8 -d \" \"";
	my @exe=split "\n", `$rsh $ps_cmd_exe`;
	my $t=0;
	foreach (0..$size-1){
		if($cpu[$_]>40){
			$t++;
			print "\t\033[1;35;49m$node\033[m\t$exe[$_]\t\033[1;32;49m$pid[$_]\033[m\t$cpu[$_]\n";
		}
	}
	$total_node{$node}=$t;
	$total+=$t;
}
print "\n\tTotal processes of \033[1;33;49m$user\033[m: \033[1;32;49m[$total]\033[m. Of which:\n";
foreach (sort keys %total_node){
	print "\t\033[1;32;49m$total_node{$_}\033[m at node \033[1;35;49m[$_]\033[m\n" if($total_node{$_});
}

